File: bit_manipulationm/swap.cc
#include <stdio.h>

void swap(int* a, int *b) {
    if (a != b) {
        *a ^= *b;
        *b ^= *a;
        *a ^= *b;
    }
}

void swapr(int& a, int& b) {
    if (&a != &b) {
        a ^= b;
        b ^= a;
        a ^= b;
    }
}

int main() {
    int a = 123;
    int b = 432;
    printf("%d, %d\n", a, b);
    swap(&a, &b);
    printf("%d, %d\n", a, b);
    swapr(a, b);
    printf("%d, %d\n", a, b);
}

File: bit_manipulationm/swap.go
package main

import "fmt"

func swap(lv, rv *int) {
	if *lv != *rv {
		*lv ^= *rv
		*rv ^= *lv
		*lv ^= *rv
	}
}

func main() {
	a, b := 2, 4
	fmt.Println(a, b)
	swap(&a, &b)
	fmt.Println(a, b)
}
